olm/operator-registry.md

```diff
@@ -0,0 +1,324 @@
+
+### Implementation Details
+
+The goal of this enhancement is to replace the app-registry implementation in a way that is usable by operator developers and CI pipeline developers. To do this we will build an index image that is defined from a set of operator bundle images that are released independently and optimized in a way that can be productized by operator authors and pipeline developers.
+
+This implementation makes the assumption that a separate process is built that generates container images that contain individual operator bundles for each version of an operator. The intent of the index image is to aggregate these images that are out of scope of this enhancement. Once that image is built, the image itself is immutable. In order to create a new image, the tooling described here will run off cluster. Once the image is loaded onto the cluster, the only way to load new content is to pull down a new version of that image (either manually or by an automated poll).
+
+The final implementation goal of this enhancement involves this set of steps:
+
+1. Use a reference to a container image (also referred to here as the operator index) rather than app-registry namespaces to expose installable operators on a cluster
+2. Build an operator-registry database that serves the data required by OLM to drive install and UI workflows
+3. Have a way to optimize the build workflow of operator-registry databases (which currently drive OLM's workflows) from previous versions of the database plus new content (ex. new operator at new version) so that the database need not be built from scratch every time they are created.
+
+To start, let's define what the operator-registry does today. It:
+
+1. Takes a set of manifest files and outputs a database file.
+2. Takes a database file and serves a grpc api that serves content from that db
+3. Takes a configmap reference, builds a db and serves a grpc api that serves content from that db
+4. Takes an app-registry path/namespace reference, builds a db and serves a grpc api that serves content from that db
+
+Below are a set of implementation steps that will add new features to accomplish our goals:
+
+*Update the Operator Registry to insert incremental updates*
+
+Add create db, delete, and batch insert APIs to the model layer of operator-registry.
+
+Add a new set of operator registry commands to utilize those new APIs:
+
+`operator-registry create`
+    - inputs: none
+    - outputs: empty operator registry database
+    
+`operator-registry add`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add quay.io/community-operators/foo:sha256@abcd123 example.db`
+    
+`operator-registry add --batch`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add "quay.io/community-operators/foo:sha256@abcd123,quay.io/community-operators/bar:sha256@defg456" example.db`
```

Comment:
For the downstream builds, how is this command going to be run inside the image? 

I understand that we are going to wrap in tooling, but making sure that the underlying thing can be used by downstream is important. 

Specifically, I am worried about prod image. This seems that the tool will pull the image, but we would not have published the production image (registry.redhat.io) yet when building this.  Is there some other mechanism to do this?

@ralphbean

Reply:
Yes, the operator-registry will use container tooling to pull the image down, inspect the manifests, and build the database.

I'm confused about the comment:

> Specifically, I am worried about prod image. This seems that the tool will pull the image, but we would not have published the production image (registry.redhat.io) yet when building this. Is there some other mechanism to do this?

What do you mean by this? The index build would expect that the images it is being built from do already exist. Otherwise whatever CI process that decides to build a new index won't have any idea what to build. How else could you build something from something else without the first thing already existing? Or am I misunderstanding?

Reply:
So
> Yes, the operator-registry will use container tooling to pull the image down, inspect the manifests, and build the database

1. I don't think that the above will work. Brew does not have access to the internet, so when creating this image, wouldn't it need access to the internet? Would there be a way of saying use this internal registry image here but name it X as we do with OMPS on a push to application registries? 
2. inspecting the images by pulling the tarball and expanding the image into the filesystem?

> Otherwise whatever CI process that decides to build a new index won't have any idea what to build. How else could you build something from something else without the first thing already existing? Or am I misunderstanding?

I am worried that this might not have been true. That we would want to pre-stage this image with the other to be delivered prod images. We could do a separate thing but I think it would require much more CI work from the pipeline/RCM side and I wanted to make sure that everyone is aware of what we are thinking will happen here.